core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * A mutable atomic reference for the `IO` monad. This is the `IO` equivalent of
+ * a volatile `var`, augmented with atomic operations, which make it useful as a
+ * reasonably efficient (if low-level) concurrency primitive.
+ *
+ * {{{
+ * for {
+ *   ref <- Ref.make(2)
+ *   v   <- ref.update(_ + 3)
+ *   _   <- console.putStrLn("Value = " + v) // Value = 5
+ * } yield ()
+ * }}}
+ */
+trait ZRef[+EA, -A, +EB, +B] {
+
+  /**
+   * Reads the value from the `Ref`.
+   */
+  def get: IO[EB, B]
+
+  /**
+   * Writes a new value to the `Ref`, with a guarantee of immediate
+   * consistency (at some cost to performance).
+   */
+  def set(a: A): IO[EA, Unit]
+
+  /**
+   * Writes a new value to the `Ref` without providing a guarantee of
+   * immediate consistency.
+   */
+  def setAsync(a: A): IO[EA, Unit]
+
+  def updateUnify[C](ea: C => A, eb: B => C)(f: C => C): IO[EA, C]
```

Comment:
is that what you meant @jdegoes ? you said it should return a `Ref[C]` but that would seem to be the `map` combinator and would demand to create a new `AtomicReference[C]` ?